### ISSUE TITLE
references with table_name_prefix causes ActiveRecord::StatementInvalid

### ISSUE THREAD

**Issue by ileodo** [rails/rails#22989]
relative to issue #20689 https://github.com/rails/rails/issues/20689
### Summary:

rails: v4.2.4
db_adapter: mysql2

when using `t.references :user, index: true, foreign_key:true` in migration with table_name_prefix, the following fail happened. however change to `add_foreign_key` method, the problem solved.

```
Mysql2::Error: Can't create table `demo`.`#sql-18c_b` (errno: 150 "Foreign key constraint is incorrectly formed"): ALTER TABLE `app_posts` ADD CONSTRAINT `fk_rails_bbd8e0451f` 
FOREIGN KEY (`user_id`)
REFERENCES `users` (`id`)
```

**Comment by mraj-rpx** [comment]
@ileodo I am able to run the migration successfully with the same scenario and below is the migration,

```ruby
class CreateComments < ActiveRecord::Migration
  def change
    create_table :comments do |t|
      t.string :title
      t.text :body
      t.references :user, index: true, foreign_key: true

      t.timestamps null: false
    end
  end
end
```
I tried with the ruby verion is v2.2.2 and rails v4.2.4.
Please let me know if I am missing something.

**Comment by ileodo** [comment]
@mraj-rpx  have you ever try it on mysql2 db_adapter? sqlite may use FAKE foreign key.

**Comment by mraj-rpx** [comment]
@ileodo I have tried the mysql2 adapter only.

**Comment by ebosantos** [comment]
I have the same issue. Ruby 2.3.0, Rails 2.4.5.1

**Comment by maclover7** [comment]
Can you create an executable test script using one of the templates [here](https://github.com/rails/rails/tree/master/guides/bug_report_templates), that would demonstrate the problem you are seeing?

**Comment by ebosantos** [comment]
I've discovered what's going on: I wasn't specifying the primary keys (in truth I was but in Rails 5 way so I could have bigint primary keys, which doesn't work). So I kept primary keys being generated by default and the foreign key creations worked. My suggestion is: Improve error messages.

**Comment by dalink** [comment]
I have the same issue.
rails: v4.2.6
ruby: 2.2.4
env: windows
gem 'activerecord-oracle_enhanced-adapter', '~> 1.6.0'
gem 'ruby-oci8', '~> 2.2.0'

I have set in application.rb
```
 config.active_record.table_name_prefix = "EDDY_"
```
```
class CreateStocks < ActiveRecord::Migration
  def change
    create_table :stocks do |t|
      t.date       :stock_date,    null: false
      t.string     :barcode,       null: false, limit: 13
      t.integer    :quantity,      null: false, limit: 2
      t.references :store,                               index: true, foreign_key: true
      t.integer    :internal_code, null: false

      t.timestamps null: false
    end
    add_foreign_key :stocks, :stores
  end
end
```
generate:
```
ActiveRecord::StatementInvalid: OCIError: ORA-00942: Table ou vue inexistante: ALTER TABLE "EDDY_stocks" ADD CONSTRAINT "FK_RAILS_F9DFFB4FE7"
FOREIGN KEY ("STORE_ID")
  REFERENCES "STORES" ("ID")
stmt.c:243:in oci8lib_220.so
```

Workaround:
```
class CreateStocks < ActiveRecord::Migration
  def change
    create_table :stocks do |t|
      t.date       :stock_date,    null: false
      t.string     :barcode,       null: false, limit: 13
      t.integer    :quantity,      null: false, limit: 2
      t.references :store,                               index: true
      t.integer    :internal_code, null: false

      t.timestamps null: false
    end
    add_foreign_key :stocks, :stores
  end
end
```
This work:

```
== 20160511074244 CreateStocks: migrating =====================================
-- create_table(:stocks)
   -> 0.1083s
-- add_foreign_key(:stocks, :stores)
   -> 0.0321s
   -> 0 rows
== 20160511074244 CreateStocks: migrated (0.1476s) ============================
```

**Comment by kamipo** [comment]
Fixed in #24614.
